Source/WebCore/dom/FullscreenManager.cpp

```diff
@@ -509,10 +509,22 @@ bool FullscreenManager::willEnterFullscreen(Element& element)
 
         ancestor->document().resolveStyle(Document::ResolveStyleType::Rebuild);
 
+        auto* renderer = ancestor->renderer();
+        auto* oldContainingBlock = renderer ? renderer->containingBlock() : nullptr;
+
         // Remove before adding, so we always add at the end of the top layer.
         if (ancestor->isInTopLayer())
             ancestor->removeFromTopLayer();
         ancestor->addToTopLayer();
+
+        // resolveStyle() will trigger layout if there are container queries, we need to invalidate
+        // the renderer again since the containing block could have since changed.
+        renderer = ancestor->renderer();
+        auto* newContainingBlock = renderer ? renderer->containingBlock() : nullptr;
+        if (oldContainingBlock != newContainingBlock) {
+            renderer->setChildNeedsLayout(MarkOnlyThis);
+            renderer->layout();
+        }
```

Comment:
If the problem is the lack of correct layout invalidation then it would be better to do in removeFromTopLayer/addToTopLayer so other call sites would get this right too. Having to do setChildNeedsLayout manually indicates that the renderer tree dirty bits are in inconsistent state, can we avoid getting into such state in the first place?

Why is synchronous layout needed here? Calling layout on a single renderer and skipping the normal layout entry points seems sketchy.